tree/ntuple/v7/inc/ROOT/RField.hxx

```diff
@@ -799,6 +799,69 @@ public:
    void CommitCluster() final;
 };
 
+/// The field for values that may or may not be present in an entry. Parent class for unique pointer field and
+/// optional field. A nullable field cannot be instantiated itself but only its descendants.
+/// The RNullableField takes care of the on-disk representation. Child classes are responsible for the in-memory
+/// representation.  The on-disk representation can be "dense" or "sparse". Dense nullable fields have a bitmask
+/// (true: item available, false: item missing) and serialize a default-constructed item for missing items.
+/// Sparse nullable fields use a (Split)Index[64|32] column to point to the available items.
+/// By default, items whose size is smaller or equal to 4 bytes (size of (Split)Index32 column element) are stored
+/// densely.
+class RNullableField : public Detail::RFieldBase {
+   /// For a dense nullable field, used to write a default-constructed item for missing ones.
+   Detail::RFieldValue fDefaultItemValue;
+   /// For a sparse nullable field, the number of written non-null items in this cluster
+   ClusterSize_t fNWritten{0};
+
+protected:
+   const Detail::RFieldBase::RColumnRepresentations &GetColumnRepresentations() const final;
+   void GenerateColumnsImpl() final;
+   void GenerateColumnsImpl(const RNTupleDescriptor &) final;
+
+   std::size_t AppendNull();
+   std::size_t AppendValue(const Detail::RFieldValue &value);
+   /// Given the index of the nullable field, returns the corresponding global index of the subfield or,
+   /// if it is null, returns kInvalidClusterIndex
+   RClusterIndex GetItemIndex(NTupleSize_t globalIndex);
+
+   RNullableField(std::string_view fieldName, std::string_view typeName, std::unique_ptr<Detail::RFieldBase> itemField);
+
+public:
+   RNullableField(RNullableField &&other) = default;
+   RNullableField &operator=(RNullableField &&other) = default;
+   ~RNullableField() override;
+
+   bool IsDense() const { return GetColumnRepresentative() == ColumnRepresentation_t({EColumnType::kBit}); }
```

Comment:
The compiler assembler output for this (given that `ColumnRepresentation_t` is essentially a `std::vector<EColumnType>`), most likely involves constructing a temporary `std::vector` and a comparison.

Given that this is used in `AppendValue()` and also in `GetItemIndex()` for reading, I think it should be more efficient to do
```suggestion
   bool IsDense() const { const auto &repr = GetColumnRepresentative(); return repr.size() == 1 && repr[0] == EColumnType::kBit; }
```
(note that, most likely, we could also omit the `repr.size() == 1` check).  Alternatively, we could use an additional data member in `RNullableField`.

Reply:
Thanks, good catch!